app/javascript/components/modals/BugReportModal.tsx

```diff
@@ -0,0 +1,100 @@
+import React, { useRef, useState, useCallback, useEffect } from 'react'
+import Modal from 'react-modal'
+import { useRequest } from '../../hooks/use-request'
+import { useIsMounted } from 'use-is-mounted'
+
+enum BugReportModalStatus {
+  INITIALIZED = 'initialized',
+  SENDING = 'sending',
+  SUCCEEDED = 'succeeded',
+}
+
+const Status = ({ status }: { status: BugReportModalStatus }) => {
+  switch (status) {
+    case BugReportModalStatus.SUCCEEDED:
+      return <p>Thanks for submitting a bug report</p>
+    case BugReportModalStatus.SENDING:
+      return <p>Sending report...</p>
+    default:
+      return null
+  }
+}
+
+export const BugReportModal = ({
+  open,
+  onClose,
+}: {
+  open: boolean
+  onClose: () => void
+}): JSX.Element => {
+  const [status, setStatus] = useState(BugReportModalStatus.INITIALIZED)
+  const url = document.querySelector<HTMLMetaElement>(
+    'meta[name="bug-reports-url"]'
+  )?.content
+  const controllerRef = useRef<AbortController | undefined>(
+    new AbortController()
+  )
+  const isMountedRef = useIsMounted()
+  const handleSubmit = useCallback(
+    (e) => {
+      e.preventDefault()
+      if (!url) {
+        return
+      }
+
+      setStatus(BugReportModalStatus.SENDING)
+
+      controllerRef.current?.abort()
+      controllerRef.current = undefined
+
+      const [request, cancel] = useRequest(
+        url,
+        JSON.stringify({ bug_report: { body: e.target.elements.body.value } }),
+        'POST'
+      )
+
+      controllerRef.current = cancel
+
+      return request
+        .then((json: any) => {
+          if (!isMountedRef.current) {
+            throw new Error('Component not mounted')
+          }
+
+          setStatus(BugReportModalStatus.SUCCEEDED)
+
+          return json
+        })
+        .catch((err) => {
+          if (err.message === 'Component not mounted') {
```

Comment:
I know this is unlikely to change, but checking an error message and conditionally executing code based on that message feels slightly brittle to me. Will this only match when the error is thrown in the `then` clause above?

Reply:
I agree. I will refactor this one.

Reply:
Use a custom error, and then you can `instanceof`